lapack: add Dlatdf

soypat opened the pull request:
Please take a look.

@vladimir-ch  There is something real fishy here. Like you said:
> This says that the slice rhs given to Dlaswp is too short. Based on the call above, rhs should be an (n-1)x1 matrix (that is, a column vector) with the stride ldz.
Though this is not the case! If rhs has stride ldz then this routine operates on non-acessible data. So rhs probably has stride 1 though I can't for the life of me figure out how those `Dlaswp` calls work.

vladimir-ch commented on lapack/gonum/dlatdf.go:
```diff
@@ -0,0 +1,161 @@
+	case ldz < max(1, n):
+		panic(badLdZ)
+	case len(rhs) < (n-1)*ldz:
+		// rhs is expected to have stride ldz
```
The Fortran code might make it look like that but `rhs` is just a column n-vector represented by a continuous array (increment of 1). For the Dlaswp call the Fortran code needs to represent it as a matrix and so they need a stride even though it's actually not used (there is no 2. column to which stride to). They use ldz as a placeholder, they could use n just as well. But since Gonum is row-major, our code should use 1 as the stride for rhs and xm.

vladimir-ch commented on lapack/gonum/dlatdf.go:
```diff
@@ -0,0 +1,161 @@
+		panic(badLenIpiv)
+	case len(jpiv) < n:
+		panic(badLenJpiv)
+	case len(z) > (ldz+1)*8:
```
The correct check should be that `if n > 8 then panic`. Then the length check for `z` can be the usual `if len(z) < (n-1)*ldz + n then panic`

vladimir-ch commented on lapack/gonum/dlatdf.go:
```diff
@@ -0,0 +1,161 @@
+		bi.Dcopy(n, work[n:], 1, xm, 1)
+
+		// Compute rhs.
+		impl.Dlaswp(1, xm, ldz, 0, n-2, ipiv[:n-1], -1)
```
```suggestion
		impl.Dlaswp(1, xm, 1, 0, n-2, ipiv[:n-1], -1)
```

vladimir-ch commented on lapack/gonum/dlatdf.go:
```diff
@@ -0,0 +1,161 @@
+
+	// Apply permutations IPIV to RHS
+	// If ijob != 2 uses look-ahead strategy.
+	impl.Dlaswp(1, rhs, ldz, 0, n-2, ipiv[:n-1], 1)
```
```suggestion
	impl.Dlaswp(1, rhs, 1, 0, n-2, ipiv[:n-1], 1)
```

soypat commented:
@vladimir-ch Thank you for taking the time to explain what was off in the implementation. I have applied your suggestions to the code. 

Please take a look at the first error check lines of Dlatdf, I have added a few new checks. 

I have not completed the tests as I am having a hard time replicating the conditions given in `Dtgsy2` (I suspect, not sure about this either). What about testing this routine via `Dtgsy2` (which I've already implemented)?

soypat commented:
Sounds good to me. Here's the PR.
